sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -1783,7 +1783,9 @@ object functions {
    * @group math_funcs
    * @since 1.4.0
    */
-  def ceil(e: Column): Column = ceil(e, lit(0))
+  def ceil(e: Column): Column = withExpr {
+    UnresolvedFunction(Seq("ceil"), Seq(e.expr), isDistinct = false)
```

Comment:
hmm, is it just a code cleanup or it does fix a bug?

Reply:
To confirm, the problem you hit is: `ceil(input)` and `ceil(input, 0)` return the same result, but use different expressions which break some custom catalyst rules?

If that's the case, I'd suggest we also fix `CeilFloorExpressionBuilderBase` and call `buildWithOneParam` if the `scale` is 0, to make `ceil(input)` and `ceil(input, 0)` exactly the same.

Reply:
@cloud-fan I agree as the current change does not take into consideration that `ceil(input)` and `ceil(input, 0)` return different types although they mean the same thing (which could confuse end users).

Reply:
I think we can call `ceil(input)`/`floor(input)` if `scale = 0`, but we need the test case to validate the modifications.   @awdavidson @revans2 Could you please provide us the test case to be sure ? thanks.

Reply:
It is a bug.

```scala
scala> spark.range(1).selectExpr("id", "ceil(id) as one_arg_sql", "ceil(id, 0) as two_arg_sql").select(col("*"), ceil(col("id")).alias("one_arg_func"), ceil(col("id"), lit(0)).alias("two_arg_func")).printSchema
root
 |-- id: long (nullable = false)
 |-- one_arg_sql: long (nullable = true)
 |-- two_arg_sql: decimal(20,0) (nullable = true)
 |-- one_arg_func: decimal(20,0) (nullable = true)
 |-- two_arg_func: decimal(20,0) (nullable = true)
 

scala> spark.range(1).selectExpr("cast(id as double) as id").selectExpr("id", "ceil(id) as one_arg_sql", "ceil(id, 0) as two_arg_sql").select(col("*"), ceil(col("id")).alias("one_arg_func"), ceil(col("id"), lit(0)).alias("two_arg_func")).printSchema
root
 |-- id: double (nullable = false)
 |-- one_arg_sql: long (nullable = true)
 |-- two_arg_sql: decimal(30,0) (nullable = true)
 |-- one_arg_func: decimal(30,0) (nullable = true)
 |-- two_arg_func: decimal(30,0) (nullable = true) 
```

Without this patch the SQL and scala APIs produce different results and the scala API produces a result that is different from what was in Spark 3.2.

I documented this in the JIRA https://issues.apache.org/jira/browse/SPARK-38604 

After this patch the single argument version behaves like it did in 3.2 and is also consistent with the SQL API.

Reply:
I added test cases that explicitly check the result type.

From a consistency standpoint if the return type is going to depend on the scale, then the scale can only ever be a literal value. If we want to break backwards compatibility, then I would suggest that we also fix the overflow issue https://issues.apache.org/jira/browse/SPARK-28135 with double being round to a long. That technically also applies to a double being cast to a decimal type and then rounded.